File: topics/arrays/easy/0121-best-time-to-buy-and-sell-stock/solution.cpp
// 0121 · Best Time to Buy and Sell Stock
// Approach: Single pass | Time: O(n) | Space: O(1)

#include <vector>
#include <algorithm>
#include <climits>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int minPrice = INT_MAX, maxProfit = 0;
        for (int price : prices) {
            minPrice = min(minPrice, price);
            maxProfit = max(maxProfit, price - minPrice);
        }
        return maxProfit;
    }
};

File: topics/arrays/easy/0121-best-time-to-buy-and-sell-stock/solution.py
# 0121 · Best Time to Buy and Sell Stock
# Approach: Single pass | Time: O(n) | Space: O(1)

from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float('inf')
        max_profit = 0
        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)
        return max_profit
